### PR TITLE
Firmware uploading

### PR DESCRIPTION
**Pull request by JakubVanek**
The goal of this PR is to implement simplified firmware uploading. I'm not yet sure whether to expose the "full" command line API or just its subset.

Fixes https://github.com/c4ev3/ev3duder/issues/16

### CODE REVIEW

**Comment by JakubVanek** [comment]
Successfully tested (by flashing 1.09D + 1.10E):
 - Windows XP Pro 32bit via USB 2.0
 - Xubuntu 18.04.5 64bit via USB 2.0

Xubuntu 18.04.5 64bit via USB 3.0 is broken. Windows 10 Pro 64bit via USB 3.1 is broken as well. However, I believe these to be hardware issues. The problem is that ev3_read_timeout reads the packet *sent* to the brick, not the reply from the brick. The same thing is visible in Wireshark - there is a URB OUT transaction, followed by URB IN transaction containing the same exact data. However, the brick does receives the sent data because `flash exit` does work, just ev3duder cannot read the reply.

On Linux, the bug triggers immediately and it breaks all bootloader commands, including `flash info`. On Win10 Pro 64bit, the error is hidden except for when the firmware download is finishing. At that point, the same situation as on Linux happens and `FW_DOWNLOAD_DATA` fails. As it happens only at the end, the new firmware is still able to boot successfully.

I will add some extra checks to detect this bug. I don't know what is causing. First I thought this is a bug in the Linux kernel, but after reproducing it on Windows, I think it might be a timing issue on the brick. It looks as if the bootloader uses an unified receive/transmit buffer and it is asked for a reply too early.

**Comment by JakubVanek** [comment]
Adding `usleep()` between `ev3_write` and `ev3_read_timeout` for `flash info` did not have any effect, neither did duplicating the `ev3_read_timeout` statement.

**Comment by JakubVanek** [comment]
How to use:
```
ev3duder flash enter                   - Reboot from Linux to the bootloader. Another option is to
                                         long-press Escape+Enter+Right on the brick.
ev3duder flash info                    - Show hardware and EEPROM versions.
ev3duder flash install <firmware.bin>  - Install the given firmware file to the EV3. After successful finish
                                         the brick will automatically reboot.
ev3duder flash exit                    - Exit from the bootloader without installing new firmware.
```

There is currently no option to do partial flashing, the whole flash is programmed always. If the file is too short, the remaining space is set to zero bytes.

File: src/main.c [hunk 1]
```diff
@@ -518,6 +589,12 @@ int main(int argc, char *argv[])
 
 		fprintf(out, "%s (%s)\n", err ?: "-", errmsg ?: "-");
 	}
+	else if (ret == ERR_USBLOOP)
+	{
+		fputs("\nError: ev3duder received its own packet instead of a reply from the brick.\n\n", stderr);
+		fputs("If you have plugged the EV3 into a USB 3.0 port (usually red or blue),\n", stderr);
+		fputs("try plugging it into a USB 2.0 only port (usually black or white).\n", stderr);
```
Comment (a3f) [inline src/main.c:596]: Huh, that's stange. I would have thought that wouldn't matter if the brick doesn't indicate USB 3.0 support per strapping the line.

**Review by a3f — APPROVED** [review]

File: src/main.c (hunk 1 above)
Comment (JakubVanek) [reply src/main.c:596]: Yup, the USB communication still has to be 2.0, but something on one side of the channel has to contain a bug that triggers this behaviour. However, I cannot point with certainty at the host or at the device, as I don't know what is really transmitted through the bus, only what the Linux receives from its drivers.

**Pull request merged**